papermill/preprocess.py

```diff
@@ -0,0 +1,180 @@
+
+# tqdm creates 2 globals lock which raise OSException if the execution
+# environment does not have shared memory for processes, e.g. AWS Lambda
+try:
+    from tqdm import tqdm
+    no_tqdm = False
+except OSError:
+    no_tqdm = True
+
+PENDING = "pending"
+RUNNING = "running"
+COMPLETED = "completed"
+
+
+class PapermillExecutePreprocessor(ExecutePreprocessor):
+    """Module containing a preprocessor that executes the code cells
+    and updates outputs"""
+
+    # Copyright (c) IPython Development Team.
+    # Distributed under the terms of the Modified BSD License.
+
+    # TODO: Delete this wrapper when nbconvert allows for setting preprocessor
+    # hood in a more convienent manner
+    def preprocess(self, nb, resources):
+        """
+        Copied with one edit of super -> papermill_preprocessor from nbconvert
+
+        Preprocess notebook executing each code cell.
+
+        The input argument `nb` is modified in-place.
+
+        Parameters
+        ----------
+        nb : NotebookNode
+            Notebook being executed.
+        resources : dictionary
+            Additional resources used in the conversion process. For example,
+            passing ``{'metadata': {'path': run_path}}`` sets the
+            execution path to ``run_path``.
+
+        Returns
+        -------
+        nb : NotebookNode
+            The executed notebook.
+        resources : dictionary
+            Additional resources used in the conversion process.
+        """
+        path = resources.get('metadata', {}).get('path', '')
+        if path == '':
+            path = None
+
+        # clear display_id map
+        self._display_id_map = {}
+
+        # from jupyter_client.manager import start_new_kernel
+
+        def start_new_kernel(startup_timeout=60, kernel_name='python', **kwargs):
+            km = self.kernel_manager_class(kernel_name=kernel_name)
+            km.start_kernel(**kwargs)
+            kc = km.client()
+            kc.start_channels()
+            try:
+                kc.wait_for_ready(timeout=startup_timeout)
+            except RuntimeError:
+                kc.stop_channels()
+                km.shutdown_kernel()
+                raise
+
+            return km, kc
+
+        kernel_name = nb.metadata.get('kernelspec', {}).get('name', 'python')
+        if self.kernel_name:
+            kernel_name = self.kernel_name
+        self.log.info("Executing notebook with kernel: %s" % kernel_name)
+        self.km, self.kc = start_new_kernel(
+            startup_timeout=self.startup_timeout,
+            kernel_name=kernel_name,
+            extra_arguments=self.extra_arguments,
+            cwd=path)
+        self.kc.allow_stdin = False
+        self.nb = nb
+
+        try:
+            nb, resources = self.papermill_preprocess(nb, resources)
+        finally:
+            self.kc.stop_channels()
+            self.km.shutdown_kernel(now=self.shutdown_kernel == 'immediate')
+
+        delattr(self, 'nb')
```

Comment:
Using `delattr` makes me take note as it is so rare. Should there be a comment explaining why we assign it on line 90 and then get rid of it again here. Seemingly never having done anything with it?

Reply:
That function was moved without any changes. I can clean it up in the move process.

Reply:
Thanks! Tidying as we go along :)